### ISSUE TITLE
Discrepancy Between Graph Outputs and Figure 7 in Reference PDF

### ISSUE THREAD

**Issue by babcockt18** [babcockt18/On-Nonlinearities-in-QML-Paper-Implementation#2]
## Issue Description
I've been working on replicating the results presented in Figure 7 of the referenced PDF within my Jupyter notebook using PennyLane. However, I'm encountering an issue where the outputs plotted on the graph do not match the expected results as shown in the PDF.

## Steps to Reproduce
1. Run the simulation with the specified parameters and quantum circuit.
2. Plot the graph of the output results.
3. Compare the plotted graph with Figure 7 in the PDF.

## Expected Behavior
The graph should closely resemble the results shown in Figure 7 of the PDF, with similar trends and data points.

## Actual Behavior
The outputs on the graph significantly differ from those presented in Figure 7 of the PDF. The discrepancies are noticeable in terms of the shape of the curve and the data points' values.

## Troubleshooting Steps Taken
- I've double-checked the quantum circuit implementation to ensure it matches the description in the paper.
- I've verified the simulation parameters and made sure they align with those used in the reference.
- I've explored different measurement functions (`qml.density_matrix` and `qml.counts`) to understand their outputs better.

## Questions
- Could there be an issue with how I'm using PennyLane's measurement functions, particularly with `qml.counts`, to calculate the variance?
- Is there a known limitation or consideration when using `qml.density_matrix` with finite shots that could lead to this discrepancy?
- Are there any suggestions for alternative approaches to measure the weighted states that could provide results more in line with the reference?

Any insights or suggestions from the community would be greatly appreciated as I work towards resolving this issue.

Thank you!

![image](https://github.com/babcockt18/On-Nonlinearities-in-QML-Paper-Implementation/assets/39506616/cf37dcb9-898e-4067-a246-ecc8ac356753)

![Screenshot 2024-02-18 at 1 14 27 PM](https://github.com/babcockt18/On-Nonlinearities-in-QML-Paper-Implementation/assets/39506616/d030004e-e66f-4f9f-b547-cd3cea6f5478)

## Reference
- [Link](https://github.com/babcockt18/On-Nonlinearities-in-QML-Paper-Implementation/blob/main/QHP.pdf) to the PDF or citation of the paper being referenced.

**Comment by babcockt18** [comment]
@AbdullahKazi500 is this something you'd be interested in helping troubleshoot?? I'm down to set up some time to meet virtually to discuss if you're interested?

**Comment by AbdullahKazi500** [comment]
yes I would be interested just let me go through this @babcockt18

**Comment by babcockt18** [comment]
Awesome! lemme know if you've got questions and we can work together to answer them!

**Comment by AbdullahKazi500** [comment]
Hi @babcockt18 I Had few questions is there a time we can meet I am available  1 PM Afternoon IST Time Sunday

**Comment by babcockt18** [comment]
hey hey! @AbdullahKazi500, would be a bit late for me since I'm on EST, could we do 10:30 AM IST (midnight EST)? Also we can continue this convo over email. send me a meeting invite (zoom, etc.) for babcockt18@gmail.com

**Comment by babcockt18** [comment]
if that doesn't work we can continue to schedule for sometime during the week that works for both of us!

**Comment by AbdullahKazi500** [comment]
> ## Issue Description
> I've been working on replicating the results presented in Figure 7 of the referenced PDF within my Jupyter notebook using PennyLane. However, I'm encountering an issue where the outputs plotted on the graph do not match the expected results as shown in the PDF.
> 
> ## Steps to Reproduce
> 1. Run the simulation with the specified parameters and quantum circuit.
> 2. Plot the graph of the output results.
> 3. Compare the plotted graph with Figure 7 in the PDF.
> 
> ## Expected Behavior
> The graph should closely resemble the results shown in Figure 7 of the PDF, with similar trends and data points.
> 
> ## Actual Behavior
> The outputs on the graph significantly differ from those presented in Figure 7 of the PDF. The discrepancies are noticeable in terms of the shape of the curve and the data points' values.
> 
> ## Troubleshooting Steps Taken
> * I've double-checked the quantum circuit implementation to ensure it matches the description in the paper.
> * I've verified the simulation parameters and made sure they align with those used in the reference.
> * I've explored different measurement functions (`qml.density_matrix` and `qml.counts`) to understand their outputs better.
> 
> ## Questions
> * Could there be an issue with how I'm using PennyLane's measurement functions, particularly with `qml.counts`, to calculate the variance?
> * Is there a known limitation or consideration when using `qml.density_matrix` with finite shots that could lead to this discrepancy?
> * Are there any suggestions for alternative approaches to measure the weighted states that could provide results more in line with the reference?

Hi Trent sorry for being away but I gave this a try here are some of my thoughts on this

You are using two different measurement functions provided by PennyLane:

qml.counts: This function returns measurement outcomes and their frequencies after running the quantum circuit.
qml.density_matrix: This function returns the density matrix representation of the quantum state.
In your context, qml.counts seems appropriate since you are interested in obtaining the outcomes and their frequencies, which are necessary for calculating the expectation value and variance of the observable O.


Measurement Interpretation
we need to Ensure that the outcomes returned by qml.counts correspond correctly to your expectations. Each outcome represents a measurement result of the circuit, and their frequencies indicate how often each outcome occurred during the shots.


When using qml.counts, keep in mind that results are based on finite shots (in your case, 1000 shots per circuit run). This finite sampling can introduce statistical fluctuations, which might affect the accuracy of the expectation value and variance estimation, especially for higher powers of 𝑘
we also need to Verify that your circuit implementation matches the described QHP algorithm accurately. Small deviations in circuit construction or parameter settings can lead to significant differences in the output results.


To address the discrepancies:

Verify Circuit Implementation: Double-check the implementation of your QHP circuit against the algorithm described in the PDF. Ensure that each step (loading states, applying gates, measuring outcomes) aligns with the intended quantum operations.

Compare Expectation Values and Variance Calculation: Calculate the expectation values and variance manually for a few cases (e.g., k=2,3) and compare them against the expected results from the PDF. This step helps identify whether the issue lies in the measurement function or in the circuit implementation.
 If the discrepancies persist, explore alternative approaches for measuring the quantum states or observables. For instance, instead of directly measuring outcomes, you might consider different types of post-processing or analysis to extract the desired information more accurately.
 Print intermediate results such as the outcome frequencies (counts) and check if they match expected probabilities or trends.
also Ensure your plotting routines correctly interpret and display the data obtained from PennyLane's measurements. Adjust axis scaling or plotting parameters if necessary to better visualize the results.

**Comment by AbdullahKazi500** [comment]
> if that doesn't work we can continue to schedule for sometime during the week that works for both of us!

I will try looking at this when I have some time
